File: rust/src/utils/grid.rs
//! Grid utilities for 2D puzzle grids
//!
//! Provides a generic Grid type for working with 2D arrays.

use std::fmt::Display;

/// A 2D grid of values
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Grid<T> {
    data: Vec<T>,
    pub width: usize,
    pub height: usize,
}

impl<T: Clone> Grid<T> {
    /// Create a new grid with the given dimensions, filled with a default value
    pub fn new(width: usize, height: usize, default: T) -> Self {
        Self {
            data: vec![default; width * height],
            width,
            height,
        }
    }

    /// Create a grid from a 2D vector
    pub fn from_vec(data: Vec<Vec<T>>) -> Option<Self> {
        if data.is_empty() {
            return None;
        }
        let height = data.len();
        let width = data[0].len();

        // Ensure all rows have the same width
        if !data.iter().all(|row| row.len() == width) {
            return None;
        }

        let flat: Vec<T> = data.into_iter().flatten().collect();
        Some(Self {
            data: flat,
            width,
            height,
        })
    }

    /// Get the value at (x, y) coordinates
    pub fn get(&self, x: usize, y: usize) -> Option<&T> {
        if x < self.width && y < self.height {
            Some(&self.data[y * self.width + x])
        } else {
            None
        }
    }

    /// Get a mutable reference to the value at (x, y) coordinates
    pub fn get_mut(&mut self, x: usize, y: usize) -> Option<&mut T> {
        if x < self.width && y < self.height {
            Some(&mut self.data[y * self.width + x])
        } else {
            None
        }
    }

    /// Set the value at (x, y) coordinates
    pub fn set(&mut self, x: usize, y: usize, value: T) -> bool {
        if x < self.width && y < self.height {
            self.data[y * self.width + x] = value;
            true
        } else {
            false
        }
    }

    /// Get the 4 cardinal neighbors (N, S, E, W) of a position
    pub fn neighbors(&self, x: usize, y: usize) -> Vec<(usize, usize)> {
        let mut result = Vec::with_capacity(4);
        
        if y > 0 {
            result.push((x, y - 1)); // North
        }
        if y + 1 < self.height {
            result.push((x, y + 1)); // South
        }
        if x > 0 {
            result.push((x - 1, y)); // West
        }
        if x + 1 < self.width {
            result.push((x + 1, y)); // East
        }
        
        result
    }

    /// Get all 8 neighbors including diagonals
    pub fn neighbors_diagonal(&self, x: usize, y: usize) -> Vec<(usize, usize)> {
        let mut result = Vec::with_capacity(8);
        
        for dy in -1i32..=1 {
            for dx in -1i32..=1 {
                if dx == 0 && dy == 0 {
                    continue;
                }
                let nx = x as i32 + dx;
                let ny = y as i32 + dy;
                if nx >= 0 && ny >= 0 && (nx as usize) < self.width && (ny as usize) < self.height {
                    result.push((nx as usize, ny as usize));
                }
            }
        }
        
        result
    }

    /// Iterate over all positions and values
    pub fn iter(&self) -> impl Iterator<Item = ((usize, usize), &T)> {
        self.data.iter().enumerate().map(move |(i, v)| {
            let x = i % self.width;
            let y = i / self.width;
            ((x, y), v)
        })
    }

    /// Find the first position matching a predicate
    pub fn find<F>(&self, predicate: F) -> Option<(usize, usize)>
    where
        F: Fn(&T) -> bool,
    {
        self.iter()
            .find(|(_, v)| predicate(v))
            .map(|(pos, _)| pos)
    }
}

impl Grid<char> {
    /// Parse a grid from a multi-line string
    pub fn from_str(input: &str) -> Option<Self> {
        let data: Vec<Vec<char>> = input
            .lines()
            .filter(|line| !line.is_empty())
            .map(|line| line.chars().collect())
            .collect();
        Self::from_vec(data)
    }
}

impl<T: Display> Display for Grid<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for y in 0..self.height {
            for x in 0..self.width {
                write!(f, "{}", self.data[y * self.width + x])?;
            }
            writeln!(f)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_grid_creation() {
        let grid: Grid<i32> = Grid::new(3, 2, 0);
        assert_eq!(grid.width, 3);
        assert_eq!(grid.height, 2);
        assert_eq!(grid.get(0, 0), Some(&0));
    }

    #[test]
    fn test_grid_neighbors() {
        let grid: Grid<i32> = Grid::new(3, 3, 0);
        let neighbors = grid.neighbors(1, 1);
        assert_eq!(neighbors.len(), 4);
    }
}

File: python/src/aoc2025/utils/grid.py
"""Grid utilities for 2D puzzle grids.

Provides a Grid class for working with 2D arrays.
"""

from typing import Callable, Generic, Iterator, TypeVar

T = TypeVar("T")


class Grid(Generic[T]):
    """A 2D grid of values."""

    def __init__(self, data: list[list[T]]):
        """Create a grid from a 2D list.

        Args:
            data: 2D list where data[y][x] is the value at position (x, y)
        """
        self.data = data
        self.height = len(data)
        self.width = len(data[0]) if data else 0

    @classmethod
    def from_string(cls, input_text: str) -> "Grid[str]":
        """Parse a grid from a multi-line string.

        Each character becomes a cell.
        """
        data = [list(line) for line in input_text.strip().split("\n")]
        return cls(data)

    @classmethod
    def filled(cls, width: int, height: int, value: T) -> "Grid[T]":
        """Create a grid filled with a default value."""
        data = [[value for _ in range(width)] for _ in range(height)]
        return cls(data)

    def get(self, x: int, y: int) -> T | None:
        """Get the value at (x, y), or None if out of bounds."""
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.data[y][x]
        return None

    def set(self, x: int, y: int, value: T) -> bool:
        """Set the value at (x, y). Returns True if successful."""
        if 0 <= x < self.width and 0 <= y < self.height:
            self.data[y][x] = value
            return True
        return False

    def neighbors(self, x: int, y: int) -> list[tuple[int, int]]:
        """Get the 4 cardinal neighbors (N, S, E, W) of a position."""
        result = []
        for dx, dy in [(0, -1), (0, 1), (-1, 0), (1, 0)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < self.width and 0 <= ny < self.height:
                result.append((nx, ny))
        return result

    def neighbors_diagonal(self, x: int, y: int) -> list[tuple[int, int]]:
        """Get all 8 neighbors including diagonals."""
        result = []
        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                if dx == 0 and dy == 0:
                    continue
                nx, ny = x + dx, y + dy
                if 0 <= nx < self.width and 0 <= ny < self.height:
                    result.append((nx, ny))
        return result

    def iter(self) -> Iterator[tuple[tuple[int, int], T]]:
        """Iterate over all positions and values."""
        for y in range(self.height):
            for x in range(self.width):
                yield (x, y), self.data[y][x]

    def find(self, value: T) -> tuple[int, int] | None:
        """Find the first position with the given value."""
        for (x, y), v in self.iter():
            if v == value:
                return (x, y)
        return None

    def find_all(self, value: T) -> list[tuple[int, int]]:
        """Find all positions with the given value."""
        return [(x, y) for (x, y), v in self.iter() if v == value]

    def find_where(self, predicate: Callable[[T], bool]) -> tuple[int, int] | None:
        """Find the first position where predicate returns True."""
        for (x, y), v in self.iter():
            if predicate(v):
                return (x, y)
        return None

    def rotate_right(self) -> "Grid[T]":
        """Rotate the grid 90 degrees clockwise.

        Returns a new Grid with the rotation applied.
        """
        new_data = [
            [self.data[self.height - 1 - y][x] for y in range(self.height)]
            for x in range(self.width)
        ]
        return Grid(new_data)

    def rotate_left(self) -> "Grid[T]":
        """Rotate the grid 90 degrees counter-clockwise.

        Returns a new Grid with the rotation applied.
        """
        new_data = [
            [self.data[y][self.width - 1 - x] for y in range(self.height)]
            for x in range(self.width)
        ]
        return Grid(new_data)

    def row(self, y: int) -> list[T] | None:
        """Get an entire row by index.

        Args:
            y: The row index (0-indexed from top)

        Returns:
            A copy of the row as a list, or None if out of bounds.
        """
        if 0 <= y < self.height:
            return self.data[y][:]
        return None

    def __str__(self) -> str:
        """Convert grid to string for display."""
        return "\n".join("".join(str(cell) for cell in row) for row in self.data)

    def __repr__(self) -> str:
        return f"Grid({self.width}x{self.height})"
